service/history/replication/stream_sender.go

```diff
@@ -513,34 +515,48 @@ Loop:
 			priority != s.getTaskPriority(item) { // case: skip task with different priority than this loop
 			continue Loop
 		}
-		task, err := s.taskConverter.Convert(item)
-		if err != nil {
-			return err
-		}
-		if task == nil {
-			continue Loop
-		}
-		task.Priority = priority
-		s.logger.Debug("StreamSender send replication task", tag.TaskID(task.SourceTaskId))
-		if err := s.sendToStream(&historyservice.StreamWorkflowReplicationMessagesResponse{
-			Attributes: &historyservice.StreamWorkflowReplicationMessagesResponse_Messages{
-				Messages: &replicationspb.WorkflowReplicationMessages{
-					ReplicationTasks:           []*replicationspb.ReplicationTask{task},
-					ExclusiveHighWatermark:     task.SourceTaskId + 1,
-					ExclusiveHighWatermarkTime: task.VisibilityTime,
-					Priority:                   priority,
+
+		operation := func() error {
+			task, err := s.taskConverter.Convert(item)
+			if err != nil {
+				return err
+			}
+			if task == nil {
+				return nil
+			}
+			task.Priority = priority
+			s.logger.Debug("StreamSender send replication task", tag.TaskID(task.SourceTaskId))
+			if err := s.sendToStream(&historyservice.StreamWorkflowReplicationMessagesResponse{
+				Attributes: &historyservice.StreamWorkflowReplicationMessagesResponse_Messages{
+					Messages: &replicationspb.WorkflowReplicationMessages{
+						ReplicationTasks:           []*replicationspb.ReplicationTask{task},
+						ExclusiveHighWatermark:     task.SourceTaskId + 1,
+						ExclusiveHighWatermarkTime: task.VisibilityTime,
+						Priority:                   priority,
+					},
 				},
-			},
-		}); err != nil {
+			}); err != nil {
+				return err
+			}
+			skipCount = 0
+			metrics.ReplicationTasksSend.With(s.metrics).Record(
+				int64(1),
+				metrics.FromClusterIDTag(s.serverShardKey.ClusterID),
+				metrics.ToClusterIDTag(s.clientShardKey.ClusterID),
+				metrics.OperationTag(TaskOperationTag(task)),
+			)
+			return nil
+		}
+
+		retryPolicy := backoff.NewExponentialRetryPolicy(1 * time.Second).
+			WithBackoffCoefficient(1.2).
+			WithMaximumInterval(3 * time.Second).
+			WithMaximumAttempts(80).
+			WithExpirationInterval(3 * time.Minute)
+
+		if err := backoff.ThrottleRetry(operation, retryPolicy, IsRetryableError); err != nil {
```

Comment:
nit: Shall we send current watermark info to receiver when sender starts to retry?

Reply:
Good point. However, I don't see an easy way to implement this sending before retrying and I think the risk is low if we don't do that:
The only case I think could cause some issue is when the stream just connected, if sender stuck at retry for the first task it trying to send, receiver will not receive anything, and no ACK message will be sent back. Currently the system already behave like this, and retry is relying on restart the eventloop/reconnect stream.